### DISCUSSION TITLE
sqlalchemy 'relation does not exist' after flask migrate

### DISCUSSION THREAD

**Discussion by andre-qumulo** [miguelgrinberg/Flask-Migrate#544] [Q&A] [answered]
The detailed write up is here: https://stackoverflow.com/questions/78481263/sqlalchemy-relation-does-not-exist-after-flask-migrate

but the short version is that I'm upgrading a Flask-SqlAlchemy app from SqlAlchemy 1.3 to SqlAlchemy 2.0.  I have a unit test that walks through the migrations and am finding that when I do a `db.session.execute(text('INSERT INTO ... ` on a table that is dropped and then added in the migration then I get an error from psychopg2 that the relation does not exist.

I am trying to understand if this is a Flask-migrate related issue or if there's somewhere else I should look.

Our flask migration code looks like:
```
    @contextlib.contextmanager
    def app_context_at_revision(self, revision_title: str) -> Iterator[None]:
        revision_hash = revision_title.split('_')[0]

        with self.app.app_context():
            flask_migrate.upgrade(directory=MIGRATION_DIR, revision=revision_hash)

        with self.app.app_context():
            yield
```
and we do a `migrate.init_app(app, db)` elsewhere in the code.

I can see the table create in the output (detailed in the stack overflow ticket) but it's like it's ignoring the migration entirely.

**Comment by miguelgrinberg** [comment]
Does the table exist in your database or not? It is not clear from your question if the problem is that the table does not get created, or that the table does exist but somehow the SQL query isn't seeing it.

**Reply by andre-qumulo** [reply]
That's the part that confuses me the most.  This is an in-memory test that's running the migrations on a temporary file system so I'm not quite sure how to test that the table exists or not outside of the method the test is using already.  Do you have a suggestion on what to add to the test code to check it post migration?

**Reply by miguelgrinberg** [reply]
I'm not sure in-memory databases would work for this. You would need all accesses to come from the same database connection, which is hard to control when you have your tests and Alembic both connecting to it. I suggest you switch to a file based database, maybe that'll help.

**Reply by andre-qumulo** [reply]
So we are using `pg_ctl` to create a DB on an in-memory python filesystem.  We are then using psychopg2 to create connections to it.  I am wondering if it has something to do with the env.py file and an interaction between flask-migrate and alembic.  Do you have an example file for me to refer to somewhere to check for something obvious?

**Reply by andre-qumulo** [reply]
I was able to get into the DB after running `flask_migrate.upgrade` that should have upgraded to several revisions.  It is showing no relations.

**Reply by andre-qumulo** [reply]
My apologies for the confusion, I've been learning the system a bit better as I've been trying to debug this.  

I've been comparing the behavior of the code to the previous one and am finding that the migration code is running in offline mode where as previously it was online.

I updated the test to run the upgrades as:
```
flask_migrate.upgrade(
    directory=MIGRATION_DIR, revision=self.initial_migration, sql=True
)
```

But when I put breakpoints in the `env.py` file then `context.is_offline_mode()` is True.  I thought passing `sql=True` would make `context.is_offline_mode()` False.  Any thoughts on why this is occurring?

**Reply by miguelgrinberg** [reply]
sql=true means that you want alembic to just output SQL code for the migrations, without actually running anything. This is what Alembic calls offline mode. The default mode with sql=false runs the migrations directly on the configured database.

**Reply by andre-qumulo** [reply, marked as answer]
Ah, thank you.  I was getting confused.  I replaced my env.py file with the one from this repo, removed `sql=True`, and any other references to running the migration as sql and now it seems to work...
